Automattic/wp-calypso
# Add actions, reducer, selectors, for theme update

In preparation to introduce theme update triggers in persistent notices, this PR introduces everything needed for tree state handling.

#### Testing

Tests are passing but you might want to double check them.

To test that this actually updates a theme:

1. edit a theme in your test site and downgrade its version to force an "update available". I used the theme Dara and downgraded it to 1.2.0.
2. patch `wp-calypso/client/my-sites/stats/activity-log-tasklist/index.jsx` adding the lines with + at the beginning:
```
 import { recordTracksEvent, withAnalytics } from 'state/analytics/actions';
 import { navigate } from 'state/ui/actions';
+ import { updateTheme } from 'state/themes/actions';


        render() {
+               return <a onClick={ this.props.updateSingleTheme }>Update theme</a>;
                const pluginsToUpdate = Object.values(
                        omit( this.props.pluginsWithUpdate, this.state.dismissedPlugins )

 const mapDispatchToProps = ( dispatch, { siteId } ) => ( {
+      updateSingleTheme: () => dispatch( updateTheme( 'dara-wpcom', siteId ) ),
        updateSinglePlugin: plugin => dispatch( updatePlugin( siteId, plugin ) ),
        showErrorNotice: ( error, options ) => dispatch( errorNotice( error, options ) ),
```
Patching this file as above will render this
<img width="288" alt="captura de pantalla 2018-05-08 a la s 13 12 58" src="https://user-images.githubusercontent.com/1041600/39769564-ed474d1a-52c2-11e8-8f74-c8aeaeedf768.png">

Verify that the actions are fired
<img width="288" alt="captura de pantalla 2018-05-08 a la s 13 15 09" src="https://user-images.githubusercontent.com/1041600/39769607-036f0ec0-52c3-11e8-93a8-76ebfab8b680.png">

Optionally verify that when THEME_UPDATE is fired, there is a flag set pointing that the theme is being updated, and when THEME_UPDATE_SUCCESS  is fired, the flag is set to false
<img width="300" alt="captura de pantalla 2018-05-08 a la s 13 05 05" src="https://user-images.githubusercontent.com/1041600/39769547-e46917e6-52c2-11e8-898c-9ff795135e31.png">
<img width="315" alt="captura de pantalla 2018-05-08 a la s 13 26 13" src="https://user-images.githubusercontent.com/1041600/39769700-4fd51c50-52c3-11e8-9271-7ce16a0c667b.png">

Finally, verify in your site that the theme was updated.
